### ISSUE TITLE
GPU underutilised in Actor Critic (A2C)  Stable Baselines3 implementation

### ISSUE THREAD

**Issue by AdritaBarari** [DLR-RM/stable-baselines3#350]
Hi , I am trying to use A2C of StablesBaselines3  for training an agent on my custom environment. My problem is that my GPU Utilization is very less (around 10 % only) while my CPU utilization has hit the ceiling. Because of this the training is very very slow.  I have tried the following things as per this discussion thread addressed by @araffin 
[https://github.com/hill-a/stable-baselines/issues/308]
 - Expanding the Actor and Critic architecture to a three layer neural network having 256, 256 and 128 neurons respectively. The GPU utilization did increase after that but it was only marginal (increased from 10 % to 15 %) as in [this ](https://github.com/hill-a/stable-baselines/issues/308#issuecomment-489587027)suggestion
 - changed device argument of A2C method to 'cuda' from the default which is 'auto' - No improvement
It  is not entirely clear what worked for the user in the above mentioned discussion.
My A2C is intialized as follows:
```
import gym
from stable_baselines3 import A2C
policy_kwargs = dict(activation_fn=th.nn.ReLU,
                     net_arch=[dict(pi=[256, 256, 128], vf=[256, 256, 128])])
model = A2C("MlpPolicy", env, verbose=1,  learning_rate=linear_schedule(0.001),\
 n_steps=50, gamma=0.8, gae_lambda=1.0, ent_coef=0.2, vf_coef=0.2, \
tensorboard_log=logdir,  policy_kwargs=policy_kwargs, seed=50, device='cuda' )
model.learn(total_timesteps=50000)
```

GPU configurations are:
- Cuda version - 9.0
- GPU - NVIDIA RTX 2080i
- GPU mem - 11GB (out of that only about 1600MB is being used)

CPU configurations
- Threads per Core - 2
- Core per socket - 8
- Intel i9 core 9900 K @ 3.6 Ghz
- CPU Mem - 32GB

SW versions
- Torch - 1.6.0
- Tensorflow GPU - 2.4.1  (although I know SB3 implementation uses Torch)
- Python 3.7 on anaconda environment

Kindly advise!

**Comment by Miffyli** [comment]
I assume your custom environment has small observations/actions (1D vectors of few hundred items). In this case running code on pure CPU (`device=cpu`) can be faster as calling CUDA operations becomes a bottleneck at small scales. Try using the original network size and CPU device and see if that improves things.  Also try setting threads to one or two with `torch.set_num_threads(1)`, which can also speed up things.

If that does not work, there is not much there can be done as the bottleneck comes from synchronous nature of A2C (see e.g. [A3C paper](https://arxiv.org/abs/1602.01783)).

**Comment by AdritaBarari** [comment]
> I assume your custom environment has small observations/actions (1D vectors of few hundred items). In this case running code on pure CPU (`device=cpu`) can be faster as calling CUDA operations becomes a bottleneck at small scales.
Yes, you are right in your assumption about the environment, but very soon I will have to scale it up. I have to try it out on a bigger observation space , which will go up to order of`~ 10 ^4. What do I do in that case ? How do I leverage the power of the GPU then ?
 > 
> If that does not work, there is not much there can be done as the bottleneck comes from synchronous nature of A2C (see e.g. [A3C paper](https://arxiv.org/abs/1602.01783)).

In that case, will switching to stable baselines **version 2** help ? since they have a A3C (asynchronous) implementation which SB3 does not have yet ?

**Comment by araffin** [comment]
Hello,
I'm not sure what to add compared to what @Miffyli wrote and what has been already discussed here:  https://github.com/hill-a/stable-baselines/issues/308
A2C is meant to be run mostly on cpu (in the original paper, I'm not sure if they use any GPU).

>changed device argument of A2C method to 'cuda' from the default which is 'auto' 

"auto" will select "cuda" automatically if available...

>In that case, will switching to stable baselines version 2 help ? since they have a A3C (asynchronous) implementation which SB3 does not have yet ?

SB2 only has A2C implemented too. If you have many cores, you can use PPO1 (MPI version) of SB2.
But again, this won't use your GPU too much as most of the computation will be be spent in the simulating the env, not doing gradient steps.

Also related: https://github.com/DLR-RM/stable-baselines3/issues/314
(one solution is to make your simulation run on GPU but this requires using a custom version of SB3)
